rfc/0003-policy-signing.md

```diff
@@ -0,0 +1,440 @@
+`https://token.actions.githubusercontent.com`.
+
+The `Subject`, on the other hand, is built using:
+
+  * The repository that contains the source code of the policy
+  * The full path to the Yaml file where the `cosign sign` invocation is written
+  * The branch: `main` or a `tag`
+
+This somehow unique `Subject` is nice, because it's more granular.
+However, that makes the end user experience uglier.
+
+In the previous cases, the identity of the policy author never changed: it was
+either his public key or his email address.
+
+In this case, the `Subject` keeps changing:
+
+  * It's different on a per policy basis: each policy lives inside of a different
+    repository
+  * It's different even for the very same policy:
+    * It changes based on the release: that's because the git tag is part of the URL
+    * Even keeping the same tag (like `main` - which would be horrible from a security
+      POV), the URL could change if the Yaml file that invokes `cosign sign` is renamed.
+
+## Signing Kubewarden policies
+
+We want to sign our policies inside of the GitHub Action that takes care of pushing
+them to our container registry.
+We want to sign them in keyless mode, using the GitHub Action OIDC token.
+
+This is the most secure setup we can imagine.
+
+This leads us to find a solution that solves the problem of the "mutating `Subject`".
+We want end users have a good experience but, at the same time, we don't want to weaken
+the strengths of our signatures.
+
+Given a `Subject`:
+
+```json
+{
+  "Subject": "https://github.com/flavio/policy-secure-pod-images/.github/workflows/release.yml@refs/heads/main"
+}
+```
+
+We can easily understand who is the owner of the repository that built the policy.
+That could be an individual user or an organization.
+
+We plan to allow end users to define an "allow list" of GitHub users/organizations
+that can be trusted.
+
+## Verifying policies
+
+Policy verification settings are going to be defined using a dedicated configuration
+file. This file format is going to be shared between Policy Server and kwctl.
+
+The proposed format is the following one:
+
+```yaml
+public_keys:
+  threshold: 2
+  keys:
+    alice: |
+      -----BEGIN PUBLIC KEY-----
+      MBFKHFDGHKIJH0CAQYIKoZIzj0DAQcDQgAEX0HFTtCfTtPmkx5p1RbDE6HJSGAVD
+      BVDF6SKFSF87AASUspkQsN3FO4iyWodCy5j3o0CdIJD/KJHDJFHDFIu6sA==
+      -----END PUBLIC KEY-----
+    bob: |
+      -----BEGIN PUBLIC KEY-----
+      AUYSDHKJFGSHGSHCAQYIKoZIzj0DAQcDQgAEX0HFTtCfTtPmkx5p1RbtwDE1EVzu
+      wjQs1cCRKb5Pz/yUspkQsN3FO4iyWodCy5j3o0CdIJD/1gvq98pf4IG9tA==
+      -----END PUBLIC KEY-----
+issuers:
+  "https://token.actions.githubusercontent.com":
+  - https://github.com/flavio/
+  - https://github.com/kubewarden/
+  "https://github.com/login/oauth":
+  - john@acme.org
+annotations:
+  env: prod
+  foo: bar
+```
+
+The file is composed by three sections, which we will analyze in the next
+sections.
```

Comment:
I don't like to perform this kind of verification, I think it can lead to unexpected results.

For example, given the following configuration:

```yaml
issuers:
  "https://token.actions.githubusercontent.com":
  - https://github.com/flavio
```

Would lead to **trust** these kind of issuers:

* `https://github.com/flavio/psp-capabilities/.github/workflows/release.yml@refs/tags/v0.0.2`
* `https://github.com/flavio-the-hacker/psp-capabilities/.github/workflows/release.yml@refs/tags/v0.0.2`

A missing `/` inside of the setting could lead to this unexpected behaviour.

Moreover, I might be paranoid, but I think someone might come up asking for a way to provide RegExp to validate the `Subject`. I think this would be super scary from a security POV, many things could go wrong that way.

I think it would be better to provide a more strict and explicit configuration:

```yaml
issuers:
  GitHubActions:
  - owner: flavio
    repositories:
      - repo1
      - repo2
   - owner: kubewarden
```

In that case, we would trust the `Subject`s generated by:
* user `flavio`: only `repo1` and `repo2`. All the other flavio's repos would not be trusted
* org `kubewarden`: all the repos are trusted

Pros:
  * more explicit and secure
  * less chances to shoot yourself in the foot

Cons:
  * Less generic
  * Could force us to provide more ad-hoc code for other environments (GitLab, Tekton, ...)